x-pack/plugin/inference/src/main/java/org/elasticsearch/xpack/inference/external/amazonbedrock/AmazonBedrockInferenceClientCache.java

```diff
@@ -0,0 +1,94 @@
+/*
+ * Copyright Elasticsearch B.V. and/or licensed to Elasticsearch B.V. under one
+ * or more contributor license agreements. Licensed under the Elastic License
+ * 2.0; you may not use this file except in compliance with the Elastic License
+ * 2.0.
+ */
+
+package org.elasticsearch.xpack.inference.external.amazonbedrock;
+
+import com.amazonaws.http.IdleConnectionReaper;
+
+import org.elasticsearch.common.util.Maps;
+import org.elasticsearch.core.Nullable;
+import org.elasticsearch.core.TimeValue;
+import org.elasticsearch.xpack.inference.services.amazonbedrock.AmazonBedrockModel;
+
+import java.io.IOException;
+import java.util.Map;
+import java.util.function.BiFunction;
+
+import static java.util.Collections.emptyMap;
+
+public final class AmazonBedrockInferenceClientCache implements AmazonBedrockClientCache {
+
+    private final BiFunction<AmazonBedrockModel, TimeValue, AmazonBedrockBaseClient> creator;
+    private volatile Map<Integer, AmazonBedrockBaseClient> clientsCache = emptyMap();
+
+    public AmazonBedrockInferenceClientCache(BiFunction<AmazonBedrockModel, TimeValue, AmazonBedrockBaseClient> creator) {
+        this.creator = creator;
+    }
+
+    public AmazonBedrockBaseClient getOrCreateClient(AmazonBedrockModel model, @Nullable TimeValue timeout) throws IOException {
+        var returnClient = internalGetOrCreateClient(model, timeout);
+        flushExpiredClients();
+        return returnClient;
+    }
+
+    private AmazonBedrockBaseClient internalGetOrCreateClient(AmazonBedrockModel model, @Nullable TimeValue timeout) throws IOException {
+        final Integer modelHash = AmazonBedrockInferenceClient.getModelKeysAndRegionHashcode(model, timeout);
+        {
+            final AmazonBedrockBaseClient client = clientsCache.get(modelHash);
```

Comment:
Do we need to lock here with the current structure of using `Maps.copyMapWithAddedEntry`?

Reply:
I wonder if we should use a readwrite lock here instead if we're trying to avoid locking the entire map on a read.

Reply:
Good idea - I'll look into that...